datafusion/optimizer/src/reduce_cross_join.rs

```diff
@@ -0,0 +1,452 @@
+                        right: Arc::new(right_plan),
+                        schema: cross_join.schema.clone(),
+                    }))
+                }
+                else {
+                    // Keep track of join keys being pushed to Join nodes
+                    all_join_keys.extend(join_keys.clone());
+
+                    Ok(LogicalPlan::Join(Join {
+                        left: Arc::new(left_plan),
+                        right: Arc::new(right_plan),
+                        join_type: JoinType::Inner,
+                        join_constraint: JoinConstraint::On,
+                        on: join_keys,
+                        filter: None,
+                        schema: cross_join.schema.clone(),
+                        null_equals_null: false,
+                    }))
+                }
+            }
+            _ => {
+                let expr = plan.expressions();
+
+                // apply the optimization to all inputs of the plan
+                let inputs = plan.inputs();
+                let new_inputs = inputs
+                    .iter()
+                    .map(|plan| {
+                        reduce_cross_join(
+                            _optimizer,
+                            plan,
+                            possible_join_keys,
+                            all_join_keys,
+                        )
+                    })
+                    .collect::<Result<Vec<_>>>()?;
+
+                from_plan(plan, &expr, &new_inputs)
+            }
+    }
+}
+
+fn intersect(
+    accum: &mut Vec<(Column, Column)>,
+    vec1: &[(Column, Column)],
+    vec2: &[(Column, Column)],
+) -> ()  {
+
+    for x1 in vec1.iter() {
+        for x2 in vec2.iter() {
+            if x1.0 == x2.0 && x1.1 == x2.1
+                || x1.1 == x2.0 && x1.0 == x2.1
+            {
+                accum.push((x1.0.clone(), x1.1.clone()));
+            }
+        }
+    }
+
+    ()
+}
+
+/// Extract join keys from a WHERE clause
+fn extract_possible_join_keys(
+    expr: &Expr,
+    accum: &mut Vec<(Column, Column)>,
+) -> () {
+    match expr {
+        Expr::BinaryExpr { left, op, right } => match op {
+            Operator::Eq => match (left.as_ref(), right.as_ref()) {
+                (Expr::Column(l), Expr::Column(r)) => {
+                    // Ensure that we don't add the same Join keys multiple times
+                    if !(accum.contains(&(l.clone(), r.clone())) ||
+                        accum.contains(&(r.clone(), l.clone()))) {
+                        accum.push((l.clone(), r.clone()));
+                    }
+                    ()
+                }
+                _ => (),
+            },
+            Operator::And => {
+                extract_possible_join_keys(left, accum);
+                extract_possible_join_keys(right, accum)
+            },
+            // Fix for issue#78 join predicates from inside of OR expr also pulled up properly.
+            Operator::Or => {
```

Comment:
@alamb In case of OR predicates, first we get the list of predicates from left and right child of OR. And then we do intersect() of left and right. In case of` A1=B1 OR A2=B2 `-> left would have `A1=B1` and right would have `A2=B2` and so intersect() of these two would be `{} empty`. **These predicates will not be moved up.** 
Let` (A1=B1 and A2=30) OR (A1=B1 and A2=B2 and A2=40) `-> In this example, left and right of OR predicate have common predicate `(A1=B1)` and so `A1=B1` **would be considered as Join predicate.**

Reply:
@DhamoPS , when I reviewed your PR, I actually wrote several tests to determine what you're saying was true. But claims in English are difficult to verify, and don't leave automated artifacts for others to understand in the future. For these reasons, I think it would be good to add tests for these conditions in addition to the PR feedback.

Reply:
Yes, I agree with @avantgardnerio  -- I am fairly sure you are correct, but I would like to see a test that both removes any doubt as well as serves as protection against regression when / if this code is refactored in the future

Reply:
I have added more test cases which would cover more scenarios. Please review the latest changes.